keps/sig-storage/1710-selinux-relabeling/README.md

```diff
@@ -65,6 +65,13 @@
 This KEP tries to speed up the way that volumes (incl. persistent volumes) are made available to Pods on systems with SELinux in enforcing mode.
 Current way includes recursive relabeling of all files on a volume before a container can be started. This is slow for large volumes.
 
+** Status: withdrawn **
+This KEP proposes to mount filesystems with the right SELinux context using `mount -o context=system_u:system_r:container_t:s0:c309,c383` mount option directly, without recursively changing labels on the volume.
+This works well for in-tree volume plugins, that mount volumes on the host, but it's problematic in CSI driver containers.
+`/bin/mount`, running in a privileged container, needs host's `/sys/fs/selinux` and `/etc/selinux/config`, otherwise it [silently discards any SELinux related mount options](https://github.com/karelzak/util-linux/blob/master/libmount/src/context_mount.c#L292).
+Since there are so many CSI drivers around, we can't force them easily to propagate these files from the host into the driver containers.
```

Comment:
I don't understand why "doing the right thing in a CSI plugin" is not an axiom here?  CSI drivers that call mount need to ensure mount works correctly on a host.  CSI aren't allowed to be completely ignorant of a host in general.

If `mount` is the appropriate way to label, why would CSI plugins simply be treated as broken (with a CSI test) if they don't support the behavior?

Reply:
Nobody in the community actually tests with SELinux. SELinux was never required in CSI and requiring it basically breaks the expected behavior of a driver.

We could add a new capability field to CSI spec for those drivers that are able to mount with `-o context=`, but it requires CSI spec change. Kubernetes vendors that use SELinux still need to have *some* workaround for large volumes & drivers that don't implement the newly introduced capability.

Reply:
The other thing worth mentioning is - one big advantage of using `mount -o context` was it was instantaneous even if volume was being mounted for the very first time but using similar policy as fsGroup means at least on first mount we will always recursively change all the files's selinux label.

But since we are planning to perform `chcon` in same loop as where we perform `chown/chmod` and recursive permission change will anyways *always* performed on first mount, cost of performing `chcon` is amortized in recursive permission change loop.

Reply:
Longer term if we do want to add a new CSI capability, would we be able to add that behavior to the proposed K8s api changes here?

Reply:
Not with values `Always` / `OjnRootMismatch`. @tallclair suggested `Recursive` (=Always) and `Heuristic` (=try all tricks you have)

Reply:
What about a new 3rd value?

Reply:
That IMO leaks implementation details about SELinux. That third value has no meaning for fsGroup and IMO it means we can't merge fsGroup and SELinux together.

Even when we have a separate field for SELinux with 'Always', 'OnRootMismatch' and 'Mount'. How a StatefulSet author knows which value to use? They don't have any idea what the storage backend supports, they only know that the PVC may be quite big and may suffer from fsGroup / SELinux recursive changes.